### PR TITLE
[Hub Generated] Review request for Microsoft.AlertsManagement to add version preview/2023-07-12-preview

### PR DESCRIPTION
**Pull request by noneumar**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://openapihub.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/noneumar-alertsmanagement-Microsoft.AlertsManagement-2023-07-12-preview2...main/).
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

**Comment by noneumar** [comment]
I created this review instead of https://github.com/Azure/azure-rest-api-specs/pull/24854.
This branch was created from 2019-05-05-preview because of AlertsManagement.json was from this version in the default tag in the readme file.
The breaking changes were approved (non breaking).
I have Avocado errors that I don't know how to solve:
In 2019-05-05-preview the paths for the APIs were /subscriptionId/{subscription}/...
but on the stable version, they are {scope}/...
I followed the stable version and got these Avocado errors:
image
What should I do?
In addition, I have a linter error on the existing API that I didn't modify in this pr:
image
I don't want to fix this in this pr, what should I do?
thank you!!!

**Comment by konrad-jamrozik** [comment]
> I created this review instead of #24854. This branch was created from 2019-05-05-preview because of AlertsManagement.json was from this version in the default tag in the readme file. The breaking changes were approved (non breaking). I have Avocado errors that I don't know how to solve @raych1 @konrad-jamrozik can you please advise what should I do? In 2019-05-05-preview the paths for the APIs were /subscriptionId/{subscription}/... but on the stable version, they are {scope}/... I followed the stable version and got these Avocado errors: image What should I do? In addition, I have a linter error on the existing API that I didn't modify in this pr: image I don't want to fix this in this pr, what should I do? thank you!!!

Please have a look at the linked documentation: it provides somewhat elaborate guidance how to handle different kinds of failures: https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/ci-fix.md#avocado

**Comment by noneumar** [comment]
> Once it is your PR's turn in the ARM review queue, ARM reviewer will have a look. (see steps 1 and 2 in the diagram in PR description).

Regarding the Avocado I think it is a false positive, so I would like to discuss it with the ARM reviewer.

**Comment by konrad-jamrozik** [comment]
@noneumar please elaborate why do you think Avocado is a false positive.

**Comment by noneumar** [comment]
> @noneumar please elaborate why do you think Avocado is a false positive.

@konrad-jamrozik Here is the explanation:

Before my changes (I added a new private API version which is adding a new alert/enrichments API),
There was a mismatch between the last stable version and the last preview version in the following APIs which I inherited in this PR:

2019-05-05-preview: **/subscriptions/{}/**providers/Microsoft.AlertsManagement/alerts
stable 2019-03-01:  /**{scope}**/providers/Microsoft.AlertsManagement/alerts

2019-05-05-preview: /**subscriptions/{subscriptionId}**/providers/Microsoft.AlertsManagement/alerts/{alertId}
stable 2019-03-01:  /**{scope}**/providers/Microsoft.AlertsManagement/alerts/{alertId}

2019-05-05-preview: /**subscriptions/{subscriptionId}**/providers/Microsoft.AlertsManagement/alerts/{alertId}/changestate
stable 2019-03-01:  /**{scope}**/providers/Microsoft.AlertsManagement/alerts/{alertId}/changestate

2019-05-05-preview: /**subscriptions/{subscriptionId}**/providers/Microsoft.AlertsManagement/alerts/{alertId}/history
stable 2019-03-01:  /**{scope}**/providers/Microsoft.AlertsManagement/alerts/{alertId}/history

2019-05-05-preview: /**subscriptions/{subscriptionId}**/providers/Microsoft.AlertsManagement/alertsSummary
stable 2019-03-01:  /**{scope}**/providers/Microsoft.AlertsManagement/alertsSummary

In my PR I followed the last stable version with {scope}, and I'm being flagged regarding missing /subscriptions/{subscription}
where {scope} includes also /subscriptions/{subscription}

WDYT? 
thank you!!!!

**Comment by konrad-jamrozik** [comment]
> > @noneumar please elaborate why do you think Avocado is a false positive.
> 
> @konrad-jamrozik Here is the explanation:
> 
> Before my changes (I added a new private API version which is adding a new alert/enrichments API), There was a mismatch between the last stable version and the last preview version in the following APIs which I inherited in this PR:
> 
> 2019-05-05-preview: **/subscriptions/{}/providers/Microsoft.AlertsManagement/alerts stable 2019-03-01: /{scope}**/providers/Microsoft.AlertsManagement/alerts
> 
> 2019-05-05-preview: /**subscriptions/{subscriptionId}**/providers/Microsoft.AlertsManagement/alerts/{alertId} stable 2019-03-01: /**{scope}**/providers/Microsoft.AlertsManagement/alerts/{alertId}
> 
> 2019-05-05-preview: /**subscriptions/{subscriptionId}**/providers/Microsoft.AlertsManagement/alerts/{alertId}/changestate stable 2019-03-01: /**{scope}**/providers/Microsoft.AlertsManagement/alerts/{alertId}/changestate
> 
> 2019-05-05-preview: /**subscriptions/{subscriptionId}**/providers/Microsoft.AlertsManagement/alerts/{alertId}/history stable 2019-03-01: /**{scope}**/providers/Microsoft.AlertsManagement/alerts/{alertId}/history
> 
> 2019-05-05-preview: /**subscriptions/{subscriptionId}**/providers/Microsoft.AlertsManagement/alertsSummary stable 2019-03-01: /**{scope}**/providers/Microsoft.AlertsManagement/alertsSummary
> 
> In my PR I followed the last stable version with {scope}, and I'm being flagged regarding missing /subscriptions/{subscription} where {scope} includes also /subscriptions/{subscription}
> 
> WDYT? thank you!!!!

@noneumar thank you. I looked through your README / AutoRest config and I see few issues. Nevertheless I am suppressing Avocado so you are now unblocked on it.

Your README is not correctly defining the default tag, which may be confusing the tool. For correct definition see [this example](https://github.com/Azure/azure-rest-api-specs/blob/80ecb4de93f209199d37d8f5e29c452afb3a3620/specification/apicenter/resource-manager/readme.md?plain=1#L30). But in your README the "Suppressions" section is intermingled with "Basic information". See [here](https://github.com/Azure/azure-rest-api-specs/blob/80ecb4de93f209199d37d8f5e29c452afb3a3620/specification/alertsmanagement/resource-manager/readme.md?plain=1#L27-L45). This is why Avocado complains about `MULTIPLE_API_VERSION` [here](https://github.com/Azure/azure-rest-api-specs/pull/25472/checks?check_run_id=16146245864) which may cause it produce wrong errors.

As for the specifics of the paths you mentioned: we are talking about a preview from 2019. I see the newest preview mentioned in README with alerts management is from [2023-07-12-preview](https://github.com/Azure/azure-rest-api-specs/blob/80ecb4de93f209199d37d8f5e29c452afb3a3620/specification/alertsmanagement/resource-manager/readme.md#tag-package-preview-2023-07). The fact Avocado is looking at the previous preview is a problem with the tool and I am going to file a bug for that. We should never worry about any other preview than the latest one.

Generally speaking, when modifying API version spec paths, you must be aware of what the README / AutoRest config is declaring and know what is in the latest stable in preview. This is something you must consult with your team and owners of the API paths introduced in the latest preview and stable.

**Comment by noneumar** [comment]
Regarding the "Swagger LintDiff" failure:
image
I think it's a false alert since "/{scope}/providers/Microsoft.AlertsManagement/alerts" also can be used to get by subscription when the scope is subscriptions/{subId}, WDYT? thank you

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (mentat9) [inline specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:465]: >subscriptionId [](http://example.com/codeflow?start=5&length=14)

Not used any more? Note, if you put back the subscription scope, you can use the subscriptionId parameter from common types, so you still won't need this one.

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (mentat9) [inline specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:561]: >smartGroupId [](http://example.com/codeflow?start=5&length=12)

Can you add information to the description to tell clients what format this ID is (e.g. resourceId, guid, something else).

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (mentat9) [inline specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:1385]: >Resource [](http://example.com/codeflow?start=33&length=8)

Recommend getting this from v5 common types.

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (mentat9) [inline specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:1415]: >alertId [](http://example.com/codeflow?start=9&length=7)

Include information in the description about what kind of ID this is (e.g. resource ID, integer, guid, etc).

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (mentat9) [inline specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:1494]: >linkToApi [](http://example.com/codeflow?start=9&length=9)

Add details to the description, e.g. Url format (or whatever it is).

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (noneumar) [reply specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json]: done.

**Comment by noneumar** [comment]
@mentat9 According to my understanding the prettier check if failing because something is wrong in the check that is being run in the pipeline:
image
On my local machine prettier check passes.

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (noneumar) [reply specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:102]: alerts is an extension API, it can be used for all azure resources, subscriptions, management groups...
In the stable API the paths are defined with {scope).
In the last preview API this was mistakenly changed to only subscription although its wrong (this is an old change from 2019 before this RP was moved to our team).
We want to align to the correct way (scope) to capture all the options of the extension API.
This is not a breaking change to the customer, it's a fix of the specification.

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (noneumar) [reply specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:561]: It's a deprecated features which will get removed in the future.

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (noneumar) [reply specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:1415]: Added GUID.

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (noneumar) [reply specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:1494]: Added Url format.

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (mentat9) [reply specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:465]: Looks like you can/should remove subscriptionId parameter, since it's not referenced any more. Not holding ARM signoff for this, but recommended.

**Comment by mentat9** [comment]
Signing off for ARM with comments. Note before merge you will need to pass review for Go and Python SDK breaking changes. Link to SDK contacts to help with that is above in the first comment, tagged [3].

File: specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json
Comment (Alancere) [inline specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json:463]: please use "x-ms-parameter-location": "method"

Comment (noneumar) [reply specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2023-07-12-preview/AlertsManagement.json]: done

**Comment by noneumar** [comment]
@mentat9 I got approvals for breaking changes from GO and python but I still cannot merge because I have Swagger LintDiff errors
image
This is not related to my PR changes and I do not know how to continue. Can you please advise what should I do? thank you!

**Review by visingla-ms — APPROVED** [review]

**Pull request merged**
